### PR TITLE
UI: Add Property and Computed Property to Utils

### PR DESCRIPTION
**Pull request by mrtracy**
Add support for lazily computed properties to our library.

Computed properties depend on "explicit" properties, which have a similar
signature to mithril properties (a simple getter/setter function). However,
these properties also have an "Epoch()" value, which is a monotonically
increasing counter incremented whenever the property is set.

Computed properties are defined by a set of 'parent' properties and a function.
The function uses values of the parent properties to create a computed value.

The computed value is memoized whenever its value is accessed; however, if any
of the parent values have changed since the memoization (detected via the
parent's Epoch() values), the value is recomputed.

Computed properties also have an Epoch(), which is equal to the sum of its
parent Epochs. Computed properties can thus be chained together.

This commit also converts the existing "QueryCache" class to use Properties.

### CODE REVIEW

File: ui/ts/components/metrics.ts [hunk 1]
```diff
@@ -83,7 +83,9 @@ module Components {
 					// after every change, so we need to figure out the best way
 					// to signal to components like this.
 					let shouldRender: boolean = false;
-					shouldRender = shouldRender || !context.epoch || context.epoch < this.vm.query.epoch();
+					shouldRender = shouldRender ||
```
Comment (tamird) [inline ui/ts/components/metrics.ts]: what is this line wrapping >.<

Comment (mrtracy) [inline ui/ts/components/metrics.ts]: this file needs to be retabbed.

File: ui/ts/components/metrics.ts [hunk 2]
```diff
@@ -83,7 +83,9 @@ module Components {
 					// after every change, so we need to figure out the best way
 					// to signal to components like this.
 					let shouldRender: boolean = false;
```
Comment (tamird) [inline ui/ts/components/metrics.ts]: why do you assign this value and then immediately mutate it?

File: ui/ts/util/property.ts [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+   * Each property also has an Epoch() function which returns a monotonically
+   * increasing number indicating the version of the property; if the value
+   * returned by the property changes, the epoch value will be increased.
+   */
+  export interface ReadOnlyProperty<T> {
+    (): T;
+    // Epoch returns the current update epoch for the property. This is a
+    // numeric value that monotonically increases each time the property is
+    // updated.
+    Epoch(): number;
```
Comment (tamird) [inline ui/ts/util/property.ts]: i don't know the ts type hierarchy, but this seems like a very generic type. could you tighten it to integer?

Comment (mrtracy) [inline ui/ts/util/property.ts]: number is the only numeric type.

File: ui/ts/util/property.ts [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+   */
+  export function Prop<T>(initial: T): Property<T> {
+    // obj stores the value of the property.
+    let obj: T = initial;
+
+    // epoch is a simple number which is incremented whenever the prop is set or
+    // when update() is called.
+    let epoch: number = 0;
+
+    let propFn: any = function(value?: T): T {
```
Comment (tamird) [inline ui/ts/util/property.ts]: do you actually need this type annotation here? this is super strange

File: ui/ts/components/metrics.ts (hunk 1 above)
Comment (mrtracy) [inline ui/ts/components/metrics.ts]: Retabbed in a separate commit which I won't squash.

File: ui/ts/util/property.ts (hunk 4 above)
Comment (mrtracy) [inline ui/ts/util/property.ts]: It's the only way to construct a function with methods.  This is a very common javascript-ism, and typescript interfaces have explicit support for it, but there is currently no type-safe way to construct such an object in typescript.

Such is life.

File: ui/ts/components/metrics.ts [hunk 5]
```diff
@@ -9,151 +9,151 @@
...
+        drawGraph: any = (element: Element, isInitialized: boolean, context: any) => {
+          if (!isInitialized) {
+            nv.addGraph(this.chart);
+          }
+
+          // TODO(mrtracy): Update if axis changes. NVD3 unfortunately
+          // breaks mithril's assumption that everything is rendering
+          // after every change, so we need to figure out the best way
+          // to signal to components like this.
+          // let shouldRender: boolean = false;
```
Comment (BramGruneir) [inline ui/ts/components/metrics.ts]: Remove this line.

File: ui/ts/test/util.ts [hunk 6]
```diff
@@ -1,9 +1,193 @@
...
+    });
+
+    test("setting value works correctly.", () => {
+      for (let i = 1; i < 5; i++) {
+        let newVal = "new_" + i.toString();
+        assert.equal(testProp(newVal), newVal);
+        assert.equal(testProp(), newVal);
+        assert.equal(testProp.Epoch(), i);
+      }
+    });
```
Comment (BramGruneir) [inline ui/ts/test/util.ts]: Can you add a test for epochs updating?

File: ui/ts/test/util.ts [hunk 7]
```diff
@@ -1,9 +1,193 @@
...
+        assert.equal(computed(), "Leonardo_1_2, Donatello, Raphael and Michaelangelo(2)");
+      });
+
+      test("computed epoch is always sum of parent epochs.", () => {
+        for (var i = 1; i < 3; i++) {
+          parents.forEach((p: Utils.Property<string>) => {
+            p(p() + "_" + i.toString());
+            assert.equal(p.Epoch(), i);
+          });
+          assert.equal(computed.Epoch(), parents[0].Epoch() * parents.length);
```
Comment (BramGruneir) [inline ui/ts/test/util.ts]: This test might be more effective if for the first loop you set parent 1, for the 2nd you set 1 and 2, etc.

File: ui/ts/test/util.ts [hunk 8]
```diff
@@ -1,9 +1,193 @@
...
+        }
+      });
+    });
+  });
```
Comment (BramGruneir) [inline ui/ts/test/util.ts]: Can you add a test that does computed of computed of computed properties?  Just to ensure the chaining does indeed work.

File: ui/ts/test/util.ts (hunk 7 above)
Comment (mrtracy) [inline ui/ts/test/util.ts]: The epoch of every parent should be the same, this asserts that.

File: ui/ts/test/util.ts (hunk 8 above)
Comment (BramGruneir) [inline ui/ts/test/util.ts]: Ok, you convinced me.  Don't worry about it.

File: ui/ts/test/util.ts (hunk 7 above)
Comment (BramGruneir) [inline ui/ts/test/util.ts]: Yes, I get that.  I was thinking that updating them out of turn from each other might show a different result.

**Comment by BramGruneir** [comment]
LGTM

**Pull request merged**